<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>snpio.utils package &mdash; SNPio 1.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=859da57e" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=fc837d61"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="snpio.popgenstats package" href="snpio.popgenstats.html" />
    <link rel="prev" title="snpio.read_input package" href="snpio.read_input.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SNPio
              <img src="_static/snpio_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_script.html">Example script</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="snpio.html">snpio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="snpio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="snpio.filtering.html">snpio.filtering package</a></li>
<li class="toctree-l3"><a class="reference internal" href="snpio.plotting.html">snpio.plotting package</a></li>
<li class="toctree-l3"><a class="reference internal" href="snpio.io.html">snpio.io package</a></li>
<li class="toctree-l3"><a class="reference internal" href="snpio.analysis.html">snpio.analysis package</a></li>
<li class="toctree-l3"><a class="reference internal" href="snpio.read_input.html">snpio.read_input package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">snpio.utils package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-snpio.utils.misc">snpio.utils.misc module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-snpio.utils.sequence_tools">snpio.utils.sequence_tools module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-snpio.utils">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="snpio.popgenstats.html">snpio.popgenstats package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="snpio.html#module-snpio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">snpio</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="snpio.html">snpio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="snpio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="snpio.filtering.html">snpio.filtering package</a></li>
<li class="toctree-l4"><a class="reference internal" href="snpio.plotting.html">snpio.plotting package</a></li>
<li class="toctree-l4"><a class="reference internal" href="snpio.io.html">snpio.io package</a></li>
<li class="toctree-l4"><a class="reference internal" href="snpio.analysis.html">snpio.analysis package</a></li>
<li class="toctree-l4"><a class="reference internal" href="snpio.read_input.html">snpio.read_input package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">snpio.utils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="snpio.popgenstats.html">snpio.popgenstats package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="snpio.html#module-snpio">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SNPio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="snpio.html">snpio package</a></li>
      <li class="breadcrumb-item active">snpio.utils package</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/btmartin721/SNPio/blob/master/docs/source/snpio.utils.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="snpio-utils-package">
<h1>snpio.utils package<a class="headerlink" href="#snpio-utils-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-snpio.utils.misc">
<span id="snpio-utils-misc-module"></span><h2>snpio.utils.misc module<a class="headerlink" href="#module-snpio.utils.misc" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.misc.get_gt2iupac">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.misc.</span></span><span class="sig-name descname"><span class="pre">get_gt2iupac</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/misc.html#get_gt2iupac"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.misc.get_gt2iupac" title="Link to this definition"></a></dt>
<dd><p>Get a dictionary of genotype to IUPAC ambiguity codes.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.misc.get_int_iupac_dict">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.misc.</span></span><span class="sig-name descname"><span class="pre">get_int_iupac_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/misc.html#get_int_iupac_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.misc.get_int_iupac_dict" title="Link to this definition"></a></dt>
<dd><p>Get a dictionary of IUPAC ambiguity codes to integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.misc.get_iupac2gt">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.misc.</span></span><span class="sig-name descname"><span class="pre">get_iupac2gt</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/misc.html#get_iupac2gt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.misc.get_iupac2gt" title="Link to this definition"></a></dt>
<dd><p>Get a dictionary of IUPAC ambiguity codes to genotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.misc.get_onehot_dict">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.misc.</span></span><span class="sig-name descname"><span class="pre">get_onehot_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/misc.html#get_onehot_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.misc.get_onehot_dict" title="Link to this definition"></a></dt>
<dd><p>Get a dictionary of IUPAC ambiguity codes to one-hot encoded vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.misc.validate_input_type">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.misc.</span></span><span class="sig-name descname"><span class="pre">validate_input_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'array'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/misc.html#validate_input_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.misc.validate_input_type" title="Link to this definition"></a></dt>
<dd><p>Validates the input type and returns it as a specified type.</p>
<p>This function checks if the input <cite>X</cite> is a pandas DataFrame, numpy array, or a list of lists. It then converts <cite>X</cite> to the specified <cite>return_type</cite> and returns it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>pandas.DataFrame</em><em>, </em><em>numpy.ndarray</em><em>, or </em><em>List</em><em>[</em><em>List</em><em>[</em><em>int</em><em>]</em><em>]</em>) – The input data to validate and convert.</p></li>
<li><p><strong>return_type</strong> (<em>str</em><em>, </em><em>optional</em>) – The type of the returned object. Supported options include: “df” (DataFrame), “array” (numpy array), and “list”. Defaults to “array”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The input data converted to the desired return type.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame, numpy.ndarray, or List[List[int]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – If <cite>X</cite> is not of type pandas.DataFrame, numpy.ndarray, or List[List[int]].</p></li>
<li><p><strong>ValueError</strong> – If an unsupported <cite>return_type</cite> is provided. Supported types are “df”, “array”, and “list”.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">X</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">validate_input_type</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="s2">&quot;df&quot;</span><span class="p">))</span>  <span class="mi">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Outputs: a DataFrame with the data from `X`.</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-snpio.utils.sequence_tools">
<span id="snpio-utils-sequence-tools-module"></span><h2>snpio.utils.sequence_tools module<a class="headerlink" href="#module-snpio.utils.sequence_tools" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.count_alleles">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">count_alleles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#count_alleles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.count_alleles" title="Link to this definition"></a></dt>
<dd><p>Counts the total number of unique alleles in a list of genotypes.</p>
<p>This function takes a list of IUPAC or VCF-style (e.g. 0/1) genotypes and returns the total number of unique alleles. The genotypes can be in VCF or STRUCTURE-style format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – A list of IUPAC or VCF-style genotypes.</p></li>
<li><p><strong>vcf</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the genotypes are in VCF format. If False, the genotypes are in STRUCTURE-style format. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The total number of unique alleles in the list.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">l</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A/A&#39;</span><span class="p">,</span> <span class="s1">&#39;A/T&#39;</span><span class="p">,</span> <span class="s1">&#39;T/T&#39;</span><span class="p">,</span> <span class="s1">&#39;A/A&#39;</span><span class="p">,</span> <span class="s1">&#39;A/T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">count_alleles</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">vcf</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Outputs: 2</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The function removes any instances of “-9”, “-”, “N”, -9, “.”, “?” before counting the alleles.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.get_iupac_caseless">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">get_iupac_caseless</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#get_iupac_caseless"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.get_iupac_caseless" title="Link to this definition"></a></dt>
<dd><p>Split IUPAC code to two primary characters, assuming diploidy.</p>
<p>Gives all non-valid ambiguities as N.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>char</strong> (<em>str</em>) – Base to expand into diploid list.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of the two expanded alleles.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.get_major_allele">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">get_major_allele</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#get_major_allele"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.get_major_allele" title="Link to this definition"></a></dt>
<dd><p>Returns the most common alleles in a list.</p>
<p>This function takes a list of genotypes for one sample and returns the most common alleles in descending order. The alleles can be in VCF or STRUCTURE-style format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – A list of genotypes for one sample.</p></li>
<li><p><strong>num</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of elements to return. If None, all elements are returned. Defaults to None.</p></li>
<li><p><strong>vcf</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the alleles are in VCF format. If False, the alleles are in STRUCTURE-style format. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The most common alleles in descending order.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>List[str]</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">l</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A/A&#39;</span><span class="p">,</span> <span class="s1">&#39;A/T&#39;</span><span class="p">,</span> <span class="s1">&#39;T/T&#39;</span><span class="p">,</span> <span class="s1">&#39;A/A&#39;</span><span class="p">,</span> <span class="s1">&#39;A/T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">get_major_allele</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">vcf</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>  <span class="c1"># Outputs: [&#39;A&#39;, &#39;T&#39;]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The function uses the Counter class from the collections module to count the occurrences of each allele.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.get_revComp_caseless">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">get_revComp_caseless</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#get_revComp_caseless"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.get_revComp_caseless" title="Link to this definition"></a></dt>
<dd><p>Returns the reverse complement of a nucleotide, while preserving case.</p>
<p>This function takes a nucleotide character and returns its reverse complement according to the standard DNA base pairing rules. It also handles IUPAC ambiguity codes. The case of the input character is preserved in the output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>char</strong> (<em>str</em>) – The nucleotide character to be reverse complemented. Can be uppercase or lowercase.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The reverse complement of the input character, with the same case.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">char</span> <span class="o">=</span> <span class="s1">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">get_revComp_caseless</span><span class="p">(</span><span class="n">char</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Outputs: &#39;t&#39;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The function supports the following IUPAC ambiguity codes: R (A/G), Y (C/T), S (G/C), W (A/T), K (G/T), M (A/C), B (C/G/T), D (A/G/T), H (A/C/T), V (A/C/G). It also supports N (any base) and - (gap).</p></li>
</ul>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.remove_items">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">remove_items</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">all_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bad_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#remove_items"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.remove_items" title="Link to this definition"></a></dt>
<dd><p>Removes items from a list based on another list.</p>
<p>This function takes a list and removes any items that are present in a second list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>all_list</strong> (<em>List</em><em>[</em><em>Any</em><em>]</em>) – The list from which items are to be removed.</p></li>
<li><p><strong>bad_list</strong> (<em>List</em><em>[</em><em>Any</em><em>]</em>) – The list containing items to be removed from the first list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The first list with any items present in the second list removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>List[Any]</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">all_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bad_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">remove_items</span><span class="p">(</span><span class="n">all_list</span><span class="p">,</span> <span class="n">bad_list</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Outputs: [&#39;a&#39;, &#39;c&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.seqCounter">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">seqCounter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#seqCounter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.seqCounter" title="Link to this definition"></a></dt>
<dd><p>Returns a dictionary of character counts in a DNA sequence.</p>
<p>This function takes a DNA sequence and returns a dictionary where the keys are nucleotide characters and the values are their counts in the sequence. It also handles IUPAC ambiguity codes. The function is case-sensitive.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em>) – The DNA sequence to be counted.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary where the keys are nucleotide characters and the values are their counts in the sequence. The dictionary also includes a ‘VAR’ key, which is the sum of the counts of all IUPAC ambiguity codes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, int]</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;ATGCRYSWKMBDHVN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seqCounter</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
<span class="go">{&#39;A&#39;: 1, &#39;N&#39;: 1, &#39;-&#39;: 0, &#39;C&#39;: 1, &#39;G&#39;: 1, &#39;T&#39;: 1, &#39;R&#39;: 1, &#39;Y&#39;: 1, &#39;S&#39;: 1, &#39;W&#39;: 1, &#39;K&#39;: 1, &#39;M&#39;: 1, &#39;B&#39;: 1, &#39;D&#39;: 1, &#39;H&#39;: 1, &#39;V&#39;: 1, &#39;VAR&#39;: 10}</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The function supports the following IUPAC ambiguity codes: R (A/G), Y (C/T), S (G/C), W (A/T), K (G/T), M (A/C), B (C/G/T), D (A/G/T), H (A/C/T), V (A/C/G). It also supports N (any base) and - (gap).</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpio.utils.sequence_tools.simplifySeq">
<span class="sig-prename descclassname"><span class="pre">snpio.utils.sequence_tools.</span></span><span class="sig-name descname"><span class="pre">simplifySeq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpio/utils/sequence_tools.html#simplifySeq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpio.utils.sequence_tools.simplifySeq" title="Link to this definition"></a></dt>
<dd><p>Simplifies a DNA sequence by replacing all nucleotides and IUPAC ambiguity codes with asterisks.</p>
<p>This function takes a DNA sequence and returns a simplified version where all nucleotides (A, C, G, T) and IUPAC ambiguity codes (R, Y, S, W, K, M, B, D, H, V) are replaced with asterisks (*). The function is case-insensitive.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em>) – The DNA sequence to be simplified.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The simplified sequence, where all nucleotides and IUPAC ambiguity codes are replaced with asterisks (*).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;ATGCRYSWKMBDHVN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">simplifySeq</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Outputs: &#39;*************&#39;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The function supports the following IUPAC ambiguity codes: R (A/G), Y (C/T), S (G/C), W (A/T), K (G/T), M (A/C), B (C/G/T), D (A/G/T), H (A/C/T), V (A/C/G).</p>
</div>
</dd></dl>

</section>
<section id="module-snpio.utils">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-snpio.utils" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="snpio.read_input.html" class="btn btn-neutral float-left" title="snpio.read_input package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="snpio.popgenstats.html" class="btn btn-neutral float-right" title="snpio.popgenstats package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Bradley T. Martin and Tyler K. Chafin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>